### ISSUE TITLE
No warnings for using unhashable types as keys in dictionaries

### ISSUE THREAD

**Issue by danr** [microsoft/pyright#2947]
**Describe the bug**
Python does not allow unhashable types as keys in dictionaries, but pyright gives no warnings for this.

**To Reproduce**
The lines with comments below give type errors from the python interpreter, but pyright gives no error.
```python
a = {'x': 1}
b = {a: 1}     # python interpreter gives TypeError: unhashable type: 'dict'
print(a in b)  # python interpreter gives TypeError: unhashable type: 'dict'
```

**Expected behavior**
Pyright should give an error that a unhashable type is used as dictionary key.

**Comment by JelleZijlstra** [comment]
Related: python/typeshed#6244.

**Comment by erictraut** [comment]
In general, I don't like adding knowledge of internal class behaviors in pyright. Such behaviors should come from the type stubs, not be hard-coded in a type checker. This case falls in the gray area because it involves a language construct (a dictionary expression) that implicitly calls the `dict` constructor.

I'll note that mypy also does not emit an error for this condition.

For now, I'm going to decline this feature request and close the issue. We can reconsider it in the future if there is signal from other pyright users (e.g. in the form of upvotes on this issue) that this is more broadly desired.

**Comment by bijij** [comment]
I'd like to note my support for adding this functionality, pyright not supporting this due to mypy's inability caused me issues today.
